src/chording.rs

```diff
@@ -0,0 +1,288 @@
+
+/// Runtime data for a chord
+#[derive(Clone)]
+struct Chord {
+    def: &'static ChordDef,
+    in_progress: bool,
+    keys_pressed: Vec<bool, 8>,
+}
+
+impl Chord {
+    /// Create new chord from user data.
+    fn new(def: &'static ChordDef) -> Self {
+        let mut me = Self {
+            def,
+            in_progress: false,
+            keys_pressed: Vec::new(),
+        };
+        for _ in def.1 {
+            me.keys_pressed.push(false).unwrap()
+        }
+        me
+    }
+
+    fn tick(&mut self, events: &Vec<Event, 8>) {
+        for e in events {
+            for (k, _) in self
+                .def
+                .1
+                .iter()
+                .enumerate()
+                .filter(|(_, key)| **key == e.coord())
+            {
+                self.keys_pressed[k] = e.is_press();
+            }
+        }
+    }
+
+    fn contains_chord(&mut self, events: &Vec<Event, 8>) -> bool {
+        for key in self.def.1 {
+            if events
+                .iter()
+                .position(|&k| (&k.coord() == key && k.is_press()))
+                .is_none()
+            {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn handle_chord(&mut self, events: &mut Vec<Event, 8>) {
+        self.in_progress = true;
+        for key in self.def.1 {
+            if let Some(position) = events
+                .iter()
+                .position(|&k| (&k.coord() == key && k.is_press()))
+            {
+                events.swap_remove(position);
+            }
+        }
+        events
+            .push(Event::Press(self.def.0 .0, self.def.0 .1))
+            .unwrap();
+    }
+
+    fn handle_release(&mut self, events: &mut Vec<Event, 8>) {
+        if self.in_progress {
+            for key in self.def.1 {
+                if let Some(position) = events
+                    .iter()
+                    .position(|&k| (&k.coord() == key && k.is_release()))
+                {
+                    events.swap_remove(position);
+                }
+            }
+            if self.keys_pressed.iter().all(|&k| !k) {
+                events
+                    .push(Event::Release(self.def.0 .0, self.def.0 .1))
+                    .unwrap();
+                self.in_progress = false;
+            }
+        }
+    }
+}
+
+/// Two keys at once!
+pub struct Chording<const N: usize> {
+    /// Defined chords
+    chords: Vec<Chord, N>,
+}
+
+impl<const N: usize> Chording<N> {
+    /// Take the predefined chord list in.
+    pub fn new(chords: &'static [ChordDef; N]) -> Self {
+        let mut v = Vec::<Chord, N>::new();
+        for ch in chords {
+            v.push(Chord::new(ch)).ok().unwrap();
+        }
+        Self { chords: v }
+    }
+
+    /// Consolidate events and return processed results as a result.
+    pub fn tick(&mut self, mut vec: Vec<Event, 8>) -> Vec<Event, 8> {
+        for c in &mut self.chords {
+            c.tick(&vec);
```

Comment:
I think it will not work in this case:

if you have a chord defined with 2 keys, first pressing a key, and later pressing the second key will (I think) generate the chord.

Reply:
My bad, it should work. But the test is interesting.

Reply:
Yes, I believe that test works. Are you asking for an additional test or a change to a test?

Reply:
An additional test would be great, but not required.

Reply:
I believe I just added the test you were referring to. Does that resolve this?

Reply:
Yes, perfect